src/helpers/audio/worker.ts

```diff
@@ -121,25 +134,48 @@ function drawSpectrogramOntoDestinationCanvas(_generation: number): void {
   // eslint-disable-next-line @typescript-eslint/ban-ts-comment
   // @ts-ignore
   destinationSurface.commit && destinationSurface.commit();
+
+  // We only set hasBufferedImage to false after successfully drawing to the
+  // canvas so if we fail to draw (for any reason), it will automatically retry
+  // on the next update.
+  // Note that I have never seen this happen in practice, but I have included
+  // this logic as a defensive programming measure.
+  hasBufferedImage = false;
 }
 
-function setup(data: SharedBuffersWithCanvas): void {
+function setup(data: SharedBuffers): void {
   //console.log("worker:setup:", data);
   state = new WorkerState(data.state);
   sampleBuffer = new Float32Array(data.sampleBuffer);
-  destinationCanvas = data.canvas;
 
   // just use a default size - regenerate will resize it in a second
   spectrogramCanvas = new OffscreenCanvas(512, 512);
   spectrogramSurface = spectrogramCanvas.getContext("2d") as OffscreenCanvasRenderingContext2D;
 
-  destinationSurface = destinationCanvas.getContext("2d") as OffscreenCanvasRenderingContext2D;
-  destinationSurface.imageSmoothingEnabled = true;
-  destinationSurface.imageSmoothingQuality = "high";
+  // Because each pixel in the spectrogram is meaningful, we disable image
+  // smoothing so that the data is visible without any subtle modification.
+  spectrogramSurface.imageSmoothingEnabled = false;
 
   state.workerReady();
 }
 
+function setupDestinationCanvas(canvas: OffscreenCanvas): void {
+  destinationCanvas = canvas;
+
+  destinationSurface = destinationCanvas.getContext("2d", {
+    alpha: false,
+  }) as OffscreenCanvasRenderingContext2D;
+  destinationSurface.imageSmoothingEnabled = false;
+
+  // Because the worker can finish generating the spectrogram before the main
+  // thread can transfer the canvas, we may have a buffered image to draw before
+  // the canvas was available.
+  // In this case, we immediately draw the spectrogram to the canvas.
+  if (hasBufferedImage) {
+    drawSpectrogramOntoDestinationCanvas(state.generation);
+  }
+}
```

Comment:
why not just always call drawSpectrogramOntoDestinationCanvas and remove the need for a global state?

Reply:
There might be nothing to draw if the FFT hasn't been computed yet